src/http/server.cr

```diff
@@ -498,9 +501,9 @@ class HTTP::Server
     @processor.process(io, io)
   end
 
+  # This method handles exceptions raised at `Socket#accept?`.
   private def handle_exception(e : Exception)
-    e.inspect_with_backtrace STDERR
-    STDERR.flush
+    Log.debug(exception: e) { "Error while connecting a new socket" }
```

Comment:
I'd say this was more of an info-level or verbose warning personally.

I actually think the logger's default severity should be info or verbose, meaning we can add `debug` logging to the stdlib without it being on by default.

Reply:
Actually I think this has to be an error. That's exactly what it is.
BTW, why is handled in a method? it's only used once 🙂

Reply:
The purpose of the method is that sublasses can override it and alter the behaviour.

Reply:
Yes, it's an error. But it has nowhere to go from here. It just bubbles up to the current fiber's root and is printed to stderr. That's not useful at all. Do you have any better idea than sending it to `Log`? IMO this is a perfect use case.

I don't mind changing the severity to info, @RX14

Reply:
Logging it as error it does not cause a bubbling. Since it’s a condition outside of what this method and class can control neither debug nor verbose are good fit I think.

If i were checking a server and something prevents connection to be accepted I think I would rather see it as an error than a debug.

Reply:
It's not an error, or a warning, just because it's an exception. Socket errors happen all the time in the real world.

Neither is it debug info.

Verbose or info fit the best.

Reply:
No, I'm not saying the exception has to bubble up. Just saying it has to be logged as error. Please, use `Log.error` instead.
Socket errors happen all the time but in certain places we should log them. Here this error is actually quite unusual, so if it happens something more serious could be happening.

Reply:
Sorry, I misunderstood what you mean with being an error.

Reply:
I checked what others are doing: Apache and nginx both seem to emit info level log messages when TLS handshake fails.

Reply:
I don’t know the reasons for logging that as just info. Anyway here’s any exception raised at the accept call. It could be TLS handshake or anything else. Let’s log as error and if some day someone complains about too much logs we could evaluate how to discern the errors.

Reply:
Here, I'm already complaining. 🖐️ That's why I did this PR because the exceptions were overwhelming when running testssl.sh. Obviously that's a specific use case destined to cause such errors. And with logging it can at least be disabled. But generally, I want to see actual errors in my log, not a mess of connection failures.
For reference, that's the current default output when logging as error while running testssl.sh: 
https://gist.github.com/straight-shoota/59b7405edfe762a2fc6688e0e7c05f4c

Reply:
Woah! those are clients trying to connect to the server and failing to do so. I better want to see the reason in the log. Those ARE actual errors. If we want to separate (ignore) errors happening because client configuration we should do something to distinguish them first (maybe raise a different exception class from SSL classes?).

But I wont accept hide all errors by default. The default must be log any error. Period.

Reply:
I don't want these exceptions hidden, they should be logged, by default. But I don't think they are errors, they are not fatal to the application, just a single connection. These random IO exceptions happen all the time, usually cause someone just navigated away on their browser while the page is loading.

Reply:
That's the problem, you say "I don't think they are errors". In fact, we don't know and the best guess is that they are. There is a `rescue` catching ALL exceptions around the `accept` call. So unless there is a demonstration that some exceptions can be ignored, we MUST log them all. Now if you tell me there is a really common exception raised under load that we shouldn't consider as error, then let's open a separate discussion and analyze that particular case. However, I have the impression that is not the case, since I have servers running with Crystal and I didn't see logs flooded with those kind of messages.

I'll give you one example that I just tested with nginx. When the system is low on file descriptors, new connections fail to connect and nginx logs this to the `error.log`:
```
2020/04/10 23:55:55 [crit] 82077#0: accept() failed (24: Too many open files)
```

Similar error can be seen with a simple HTTP server in Go:
```
2020/04/10 23:47:20 http: Accept error: accept tcp [::]:8080: accept: too many open files; retrying in 5ms
```

Those are errors. It's something a system administrator wants to get notified ASAP.

Reply:
You're right about different errors that could be raised from that method. What you mentioned should be logged as error. But what's typically caused by an individual misbehaving should not get the same attention level.
If my server fails to accept new connections because it's file descriptors are exhausted, I want to know. But I really don't care if a client fails to connects because it can't do TLS right. That would add extra noise and could cause the really noteworthy errors disappear in the mass.

Reply:
Yes, it all depends on the error:

- Too many open files => that's a critical error happening on the server, we _can_ do something to fix it;
- Connection error => a client failed, maybe it tried HTTP on TLS port, closed the connection immediately, failed TLS handshake, or whatever, but there is _nothing_ the server can do. It's informational, at best.

Reply:
Considering how to implement this, the difficult part is discerning which category an error belongs to. That's not trivial because everything related to TLS is just `OpenSSL::SSL::Error`. We plan on changing some things here anyway, such as a generic `TLS` API and change the integration with `HTTP::Server` to do the handshake inside the client fiber.
So it's probably best to postpone solving the log severity assignment. I'll go for everything being logged as error for now and we can refine it with/after the TLS/server refactor.

Reply:
Thanks @straight-shoota. I agree 100% now.
Probably we need to subclass `OpenSSL::SSL::Error` or maybe just make the `OpenSSL::SSL::Socket::Server#accept` method to not return until a connection was made or a fatal error occur, and just log client errors from there as `info` or `debug` but with a different source.

Maybe Apache sources can give us a hint about what errors to consider here: https://github.com/apache/httpd/blob/add4e2892b4c6b35be790843aac5713302dc1d2f/modules/ssl/ssl_engine_io.c#L1410-L1453.

Reply:
I think such errors should not be handled by SSL implementation but the calling code. For an HTTPS server, it's usually okay to essentially ignore client errors. In other use cases that might need to be handled differently.
Alsoas per https://github.com/crystal-lang/crystal/issues/8108#issuecomment-531767712  `SSL::Socket::Server` needs to be removed from HTTP::Server implementation. Probably  also entirely.

Reply:
Well... the call to `SSL_accept` still has to happen somewhere, right? Aren't we just moving the problem somewhere else? But yes, having a `SSL::ClientError` subclass should be enough to be handled by servers, close everything and log as debug.

Reply:
Yes, we'd still have to sort out what's what. But it would prob be a bit easier when TCP accept and TSL accept don't happen in the same call.
(I was mainly commenting on your idea to make `SSL::Server#accept` block until it is successful and log every failure internally.)